WT-16055 Implement KEK read path mechanism

ablekhman opened the pull request:
- Read encryption key data and load into key provider
- Small refactoring of surrounding functions
- Tests for metadata parsing

ablekhman committed:
WT-16055 Implement KEK read path mechanism
- Read encryption key data and load into key provider
- Small refactoring of surrounding functions
- Tests for metadata parsing
- Bug fix: replace `strtoll` with custom implementation. `strtoll` expects null-terminated strings and can read beyond the valid buffer.

ablekhman committed:
WT-16055 Implement KEK read path mechanism
- Read encryption key data and load into key provider
- Small refactoring of surrounding functions
- Tests for metadata parsing
- Bug fix: replace `strtoll` with custom implementation. `strtoll` expects null-terminated strings and can read beyond the valid buffer.

ablekhman committed:
WT-16055 Implement KEK read path mechanism
- Read encryption key data and load into key provider
- Small refactoring of surrounding functions
- Tests for metadata parsing
- Bug fix: replace `strtoll` with custom implementation. `strtoll` expects null-terminated strings and can read beyond the valid buffer.

ablekhman committed:
WT-16055 Implement KEK read path mechanism
- Read encryption key data and load into key provider
- Small refactoring of surrounding functions
- Tests for metadata parsing
- Bug fix: replace `strtoll` with custom implementation. `strtoll` expects null-terminated strings and can read beyond the valid buffer.

ablekhman committed:
WT-16055 Implement KEK read path mechanism
- Read encryption key data and load into key provider
- Small refactoring of surrounding functions
- Tests for metadata parsing
- Bug fix: replace `strtoll` with custom implementation. `strtoll` expects null-terminated strings and can read beyond the valid buffer.

ablekhman committed:
WT-16055 Implement KEK read path mechanism
- Read encryption key data and load into key provider
- Small refactoring of surrounding functions
- Tests for metadata parsing
- Bug fix: replace `strtoll` with custom implementation. `strtoll` expects null-terminated strings and can read beyond the valid buffer.

ablekhman committed:
WT-16055 Implement KEK read path mechanism
- Read encryption key data and load into key provider
- Small refactoring of surrounding functions
- Tests for metadata parsing

ablekhman committed:
WT-16055 Implement KEK read path mechanism
- Read encryption key data and load into key provider
- Small refactoring of surrounding functions
- Tests for metadata parsing

jiechenbo commented on src/conn/conn_layered.c:
```diff
@@ -246,49 +228,152 @@ __disagg_get_meta(WT_SESSION_IMPL *session, uint64_t page_id, uint64_t lsn, WT_I
     WT_ASSERT_ALWAYS(session, page_id <= WT_DISAGG_KEY_PROVIDER_MAX_PAGE_ID,
       "Multiple key provider pages is not currently supported");
+
+    WT_RET(__disagg_get_page(session, disagg->page_log_key_provider, page_id, lsn, item));
```
Could you set the disagg->last `last_key_provider_page_lsn` here too? Writing out the key provider config depends on `last_key_provider_page_lsn` to be set. Without we would not write the key provider config into the shared turtle file.

jiechenbo commented on src/conn/conn_layered.c:
```diff
@@ -246,49 +228,152 @@ __disagg_get_meta(WT_SESSION_IMPL *session, uint64_t page_id, uint64_t lsn, WT_I
+          "Encryption key data too small: expected at least %" WT_SIZET_FMT ", got %" WT_SIZET_FMT,
+          sizeof(WT_CRYPT_HEADER), key_item->size);
+    }
+    memcpy(header, key_item->data, sizeof(WT_CRYPT_HEADER));
```
Do we need to do a memcpy here? Can't we just directly point to the header instead?

ablekhman replied:
I prefer to copy because:
1. The header will change after byte swap, so the original data stays intact.
2. The destination header is quite small and allocated on stack, so there is no much of on overhead.

jiechenbo commented on src/conn/conn_layered.c:
```diff
@@ -301,29 +386,17 @@ __disagg_put_meta(WT_SESSION_IMPL *session, uint64_t page_id, const WT_ITEM *ite
+    WT_RET(__disagg_put_page(
+      session, disagg->page_log_meta, page_id, item, disagg->last_metadata_page_lsn, lsnp));
     __wt_atomic_add_uint64_v(&disagg->num_meta_put, 1);
+
```
np: newline

jiechenbo commented on src/conn/conn_layered.c:
```diff
@@ -490,8 +559,137 @@ __wt_disagg_put_checkpoint_meta(WT_SESSION_IMPL *session, const char *checkpoint
+            __wt_config_subinit(session, &page_cfg, &cfg_value);
+            while ((ret = __wt_config_next(&page_cfg, &cfg_key, &cfg_value)) == 0) {
+                if (WT_CONFIG_LIT_MATCH("page_id", cfg_key) &&
+                  cfg_value.type == WT_CONFIG_ITEM_NUM) {
```
I wonder if `__config_check` would be a suitable replacement for the checking of config?

ablekhman replied:
I looked it up. Seems like an overkill for this check. In order to use it properly, I will have to pre-fill `WT_CONFIG_ENTRY` struct, etc..

jiechenbo commented on src/conn/conn_layered.c:
```diff
@@ -490,8 +559,137 @@ __wt_disagg_put_checkpoint_meta(WT_SESSION_IMPL *session, const char *checkpoint
+
+    /*
+     * No key provider information stored in disaggregated storage. Use empty crypt keys to let the
+     * key provider load the default key.
```
I wouldn't mention `loading the default key` here. It is up to the key provider to decide what they want to do with this information

jiechenbo commented on src/include/meta.h:
```diff
@@ -134,3 +134,17 @@ struct __wt_blkincr {
+    size_t checkpoint_len;         /* Length of checkpoint metadata string */
+    uint64_t checkpoint_timestamp; /* Checkpoint timestamp */
+
+    const char *key_provider; /* Key provider metadata string */
```
np: We have `key_provider` name attached to connection already that means the actual key provider module. Could we rename this to `key_provider_cfg`?

ablekhman replied:
I know it's a bit confusing. I keep the name to be consistent with the actual contents of the turtle file. May be the right thing to do is to rename the key itself? Something like `key_provider` -> `crypt` to follow the checkpoint pattern?
So, the turtle file will look like:
```
checkpoint=(...),
timestamp=...,
crypt=(<crypt keys metadata>)
```

jiechenbo replied:
Nah I believe key_provider sounds more clearer vses crypt. Let's keep as is 👍

jiechenbo commented on test/catch2/misc_tests/test_disagg_meta_config.cpp:
```diff
@@ -0,0 +1,214 @@
+/*-
```
I would rename test_disagg, maybe test_disagg_meta_config?

jiechenbo commented on test/catch2/misc_tests/test_disagg_meta_config.cpp:
```diff
@@ -0,0 +1,214 @@
+#include <sstream>
+#include <string_view>
+
+struct disagg_fixture {
```
Could you add an introductory comment describing what we are trying to test here?

ablekhman committed:
WT-16055 Implement KEK read path mechanism
- Read encryption key data and load into key provider
- Small refactoring of surrounding functions
- Tests for metadata parsing

jiechenbo reviewed (approved):
LGTM, i will schedule the key provider python tests to see if it passes
